concrete/jobs/generate_sitemap.php

```diff
@@ -1,243 +1,78 @@
-        $changefreq = $page->getAttribute($instances['ak_sitemap_changefreq']);
-        $priority = $page->getAttribute($instances['ak_sitemap_priority']);
-        $xmlNode = $xmlDoc->addChild('url');
-        $xmlNode->addChild(
-            'loc',
-            $instances['navigation']->getLinkToCollection($page)
-        );
-        $xmlNode->addChild('lastmod', $lastmod->format(DateTime::ATOM));
-        $xmlNode->addChild(
-            'changefreq',
-            Core::make('helper/validation/strings')->notempty($changefreq) ? $changefreq : Config::get('concrete.sitemap_xml.frequency')
-        );
-        $xmlNode->addChild(
-            'priority',
-            is_numeric($priority) ? $priority : Config::get('concrete.sitemap_xml.priority')
-        );
-        if ($instances['multilingualSections']) {
-            $thisSection = MultilingualSection::getBySectionOfSite($page);
-            if (is_object($thisSection) && !$thisSection->isError()) {
-                foreach ($instances['multilingualSections'] as $section) {
-                    $relatedPageID = $section->getTranslatedPageID($page);
-                    if ($relatedPageID) {
-                        $relatedPage = Page::getByID($relatedPageID);
-                        if (static::canIncludePageInSitemap($relatedPage, $instances)) {
-                            $xmlAltNode = $xmlNode->addChild('link', null, 'http://www.w3.org/1999/xhtml');
-                            $xmlAltNode->addAttribute('rel', 'alternate');
-                            $xmlAltNode->addAttribute('hreflang', strtolower(str_replace('_', '-', $section->getLocale())));
-                            $xmlAltNode->addAttribute('href', $instances['navigation']->getLinkToCollection($relatedPage));
-                        }
-                    }
-                }
+        $writer = $this->app->make(SitemapWriter::class);
+        $generator = $writer->getSitemapGenerator();
+        if ($generator->getSiteCanonicalUrl() === '') {
+            $request = $this->app->make(Request::class);
+            if (!$request->getHost()) {
+                throw new UserMessageException(t('Canonical URL is not set and no HTTP request to retrieve the site URL from.'));
             }
+            $generator->setCustomSiteCanonicalUrl($request->getSchemeAndHttpHost());
         }
-
-        $event = new \Symfony\Component\EventDispatcher\GenericEvent();
-        $event->setArgument('xmlNode', $xmlNode);
-        $event->setArgument('page', $page);
-        Events::dispatch('on_sitemap_xml_addingpage', $event);
-
-        if ((!empty($ret)) && ($ret < 0)) {
-            for ($i = count($xmlDoc->url) - 1; $i >= 0; --$i) {
-                if ($xmlDoc->url[$i] == $xmlNode) {
-                    unset($xmlDoc->url[$i]);
-                    break;
-                }
+        $numPages = 0;
+        $writer->generate(function (SitemapElement $data) use (&$numPages) {
```

Comment:
`function (` not sure if that is the correct coding style with the extra space?

Reply:
php-cs-fixer adds that space... Not sure which is [the rule it uses](https://mlocati.github.io/php-cs-fixer-configurator/) for that though...